File: sites/materialtools.py
"""
https://www.materialtools.com/
"""
import asyncio
from datetime import datetime
from typing import List

import aiohttp
from bs4 import BeautifulSoup

from .models import SMS, Phone, session


class Materialtools:
    async def update(self):
        site = "materialtools"

        # 首先清理该站点过往数据
        session.query(Phone).filter_by(site=site).delete()
        session.commit()
        async with aiohttp.ClientSession() as aio_session:
            # 获取页码
            async with aio_session.get(
                "https://www.materialtools.com/",
                headers={
                    "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36"
                },
            ) as resp:
                text = await resp.text()
                print(text)
            page_size = int(
                BeautifulSoup(text, "html.parser")
                .find_all("ul")[-1]
                .find_all("li")[-2]
                .text.strip()
            )
            # 获取每页的数据
            for page in range(1, page_size + 1):
                async with aio_session.get(
                    f"https://www.materialtools.com/?page={page}"
                ) as resp:
                    text = await resp.text()
                soup = BeautifulSoup(text, "html.parser")
                for div in soup.find_all(class_="sms-number-list"):
                    number_div = div.find(class_="number-list-phone_number")
                    if not number_div:  # 过滤广告列
                        continue
                    small = number_div.find("small").text.strip()
                    if small != "+86":  # 过滤非大陆手机号
                        continue
                    phone_number = number_div.find("h3").text.strip()
                    read_div = div.find(class_="sms-number-read")
                    read_link = (
                        "https://www.materialtools.com" + read_div.find("a")["href"]
                    )
                    phone = (
                        session.query(Phone)
                        .filter_by(phone_number=phone_number, site=site)
                        .first()
                    )
                    if phone:  # 判断数据重复
                        continue
                    phone = Phone()
                    phone.phone_number = phone_number
                    phone.site = site
                    phone.detail_url = read_link
                    session.add(phone)
            session.commit()

    async def detail(self, phone: Phone) -> List[SMS]:
        sms_list = []
        async with aiohttp.ClientSession() as aio_session:
            async with aio_session.get(phone.detail_url) as resp:
                text = await resp.text()
        soup = BeautifulSoup(text, "html.parser")
        div = soup.find_all(class_="sms-content-table")[1]
        for tr in div.find_all("tr"):
            items = tr.find_all("td")
            if not items:
                continue
            sms = SMS()
            sms.phone = phone
            sms.sms_id = items[0].text.strip()
            sms.send_phone = items[1].text.strip()
            sms.text = items[2].text.strip()
            sms.recv_time = datetime.strptime(
                items[3].text.strip(), "%Y-%m-%d %H:%M:%S"
            )
            # 其实好像没什么必要存起来，因为这种数据实时性比较高，每次都要重新获取的
            session.add(sms)
            sms_list.append(sms)
        phone.sms = sms_list
        session.commit()
        return sms_list

File: sites/materialtools.ts
import { Phone, SMS, Site, Browser } from './models'
import * as puppeteer from 'puppeteer'

async function filterRequest(page: puppeteer.Page) {
    await page.setRequestInterception(true);
    page.on('request', (request) => {
        const url = request.url();
        const filters = [
            'gstatic.com',
            'static.pdflibr.com',
            'googlesyndication.com',
            'googletagmanager.com',
            'google.com',
            'doubleclick.net',
            'google-analytics.com',
            'googletagservices.com',
            'googleapis.com',
        ];
        const shouldAbort = filters.some((urlPart) => url.includes(urlPart));
        if (shouldAbort) request.abort();
        else {
            request.continue();
        }
    });
}

export class Materialtools implements Site {
    name = 'materialtools'
    async list(): Promise<Array<Phone>> {
        const browser = await Browser.getBrowser()
        const page = await browser.newPage()
        await filterRequest(page)
        await page.goto('https://www.materialtools.com/')

        let data = await this.parse(page)
        const page_size = await page.evaluate(() => {
            const lis = document.querySelectorAll('ul')[3].querySelectorAll('li')
            const page_size = Number(lis[lis.length - 2].innerText.trim())
            return page_size
        })
        for (let i = 2; i <= page_size; i++) {
            await page.goto(`https://www.materialtools.com/?page=${i}`)
            data = data.concat(await this.parse(page))
        }
        await page.close()
        return data
    }
    async detail(phone: Phone): Promise<Array<SMS>> {
        const browser = await Browser.getBrowser()
        const page = await browser.newPage()
        await filterRequest(page)
        await page.goto(phone.detail_url)
        const data = await page.evaluate(() => {
            const div = document.getElementsByClassName('sms-content-table')[1]
            const tr_list = div.querySelectorAll('tr')
            const data = []
            for (let i = 1; i < tr_list.length; i++) {
                const tr = tr_list[i]
                const td_list = tr.querySelectorAll('td')
                data.push({
                    sms_id: td_list[0].innerText.trim(),
                    send_phone: td_list[1].innerText.trim(),
                    text: td_list[2].innerText.trim(),
                    recv_time: td_list[3].innerText.trim(),
                })
            }
            return data
        })
        await page.close()
        return data
    }
    async parse(page: puppeteer.Page): Promise<Array<Phone>> {
        const data = await page.evaluate((site_name) => {
            const data = []
            const sms_list = document.getElementsByClassName('sms-number-list');
            for (let i = 0; i < sms_list.length; i++) {
                const phone_number_div = sms_list[i].querySelector('.phone_number-text')
                if (!phone_number_div) {
                    continue
                }
                const small = phone_number_div.querySelector('small')
                if (!small || small.innerText.trim() !== '+86') {
                    continue
                }
                const phone_number = phone_number_div.querySelector('h3').innerText.trim()
                const detail_url = sms_list[i].querySelector('.sms-number-read').querySelector('a').href
                data.push({
                    phone_number: phone_number,
                    detail_url: detail_url,
                    site: site_name
                })
            }
            return data
        }, this.name)
        return data
    }
}